data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/worker/OpenSearchIndexPartitionCreationSupplier.java

```diff
@@ -0,0 +1,116 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Function;
+import java.util.regex.Matcher;
+import java.util.stream.Collectors;
+
+public class OpenSearchIndexPartitionCreationSupplier implements Function<Map<String, Object>, List<PartitionIdentifier>> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(OpenSearchIndexPartitionCreationSupplier.class);
+
+    private final OpenSearchSourceConfiguration openSearchSourceConfiguration;
+    private final OpenSearchClient openSearchClient;
+
+    public OpenSearchIndexPartitionCreationSupplier(final OpenSearchSourceConfiguration openSearchSourceConfiguration,
+                                                    final ClusterClientFactory clusterClientFactory) {
+        this.openSearchSourceConfiguration = openSearchSourceConfiguration;
+
+        final Object client = clusterClientFactory.getClient();
+
+        if (client instanceof OpenSearchClient) {
+            this.openSearchClient = (OpenSearchClient) client;
+        } else {
+            throw new IllegalArgumentException(String.format("ClusterClientFactory provided an invalid client object to the index partition creation supplier. " +
+                    "The client must be of type OpenSearchClient. The client passed is of class %s", client.getClass()));
+        }
+
+    }
+
+    @Override
+    public List<PartitionIdentifier> apply(final Map<String, Object> globalStateMap) {
+
+        if (Objects.nonNull(openSearchClient)) {
+            return applyForOpenSearchClient(globalStateMap);
+        }
+
+        return Collections.emptyList();
+    }
+
+    private List<PartitionIdentifier> applyForOpenSearchClient(final Map<String, Object> globalStateMap) {
+        IndicesResponse indicesResponse;
+        try {
+            indicesResponse = openSearchClient.cat().indices();
+        } catch (IOException | OpenSearchException e) {
+            LOG.error("There was an exception when calling /_cat/indices to create new index partitions", e);
+            return Collections.emptyList();
+        }
+
+        return indicesResponse.valueBody().stream()
+                .filter(this::isIndexIncludedInOneOfTheIncludePatternsAndNotExcludedInAnExcludePattern)
+                .map(indexRecord -> PartitionIdentifier.builder().withPartitionKey(indexRecord.index()).build())
+                .collect(Collectors.toList());
+    }
+
+    private boolean isIndexIncludedInOneOfTheIncludePatternsAndNotExcludedInAnExcludePattern(final IndicesRecord indicesRecord) {
+        if (Objects.isNull(indicesRecord.index())) {
+            return false;
+        }
+
+        final IndexParametersConfiguration indexParametersConfiguration = openSearchSourceConfiguration.getIndexParametersConfiguration();
```

Comment:
Take or leave it: This and other configurations values can be saved as a class variable instead of creating a new local variable for every index filter. These values are always the same.